src/Symfony/Component/Runtime/SymfonyRuntime.php

```diff
@@ -0,0 +1,198 @@
+ *  - Application, Command, InputInterface and/or OutputInterface
+ *    from "symfony/console" if the component is installed.
+ *
+ * This runtime can handle app-callables that return instances of either:
+ *  - HttpKernelInterface,
+ *  - Response,
+ *  - Application,
+ *  - Command,
+ *  - int|string|null as handled by GenericRuntime.
+ *
+ * @author Nicolas Grekas <p@tchwork.com>
+ */
+class SymfonyRuntime extends GenericRuntime
+{
+    private $request;
+    private $input;
+    private $output;
+    private $console;
+    private $command;
+    private $env;
+
+    public function __construct(array $options = [])
+    {
+        $this->env = $options['env'] ?? null;
+        $_SERVER['APP_ENV'] = $options['env'] ?? $_SERVER['APP_ENV'] ?? null;
+
+        parent::__construct($options);
+
+        if (isset($_SERVER['argv']) && null === $this->env && class_exists(ArgvInput::class)) {
+            $this->getInput();
+        }
+
+        if (($options['disable_dotenv'] ?? false) || !isset($options['project_dir']) || class_exists(MissingDotenv::class, false)) {
+            $_SERVER['APP_DEBUG'] = filter_var($options['debug'] ?? $_SERVER['APP_DEBUG'] ?? true, \FILTER_VALIDATE_BOOLEAN) ? '1' : '0';
+        } else {
+            $dotenv = new Dotenv();
+
+            if ($prodEnvs = ($options['prod_envs'] ?? null)) {
+                $dotenv->setProdEnvs((array) $dotenv);
+            }
+
+            $dotenv->bootEnv($options['project_dir'].'/.env', 'dev', $options['test_envs'] ?? ['test']);
+        }
+
+        if ($_SERVER['APP_DEBUG'] && class_exists(Debug::class)) {
+            umask(0000);
+            Debug::enable();
+        }
+    }
+
+    public function getRunner(object $application): RunnerInterface
+    {
+        if ($application instanceof HttpKernelInterface) {
+            $request = $this->request ?? $this->request = Request::createFromGlobals();
+
+            return new HttpKernelRunner($application, $request, $this);
+        }
+
+        if ($application instanceof Response) {
+            return new ResponseRunner($application);
+        }
+
+        if ($application instanceof Command) {
+            $console = $this->console ?? $this->console = new Application();
+            $console->setName($application->getName() ?: $console->getName());
+
+            if (!$application->getName() || !$console->has($application->getName())) {
+                $application->setName($_SERVER['argv'][0]);
+                $console->add($application);
+            }
+
+            $console->setDefaultCommand($application->getName(), true);
+
+            return $this->getRunner($console);
+        }
+
+        if ($application instanceof Application) {
+            if (!\in_array(\PHP_SAPI, ['cli', 'phpdbg', 'embed'], true)) {
+                echo 'Warning: The console should be invoked via the CLI version of PHP, not the '.\PHP_SAPI.' SAPI'.\PHP_EOL;
+            }
+
+            set_time_limit(0);
```

Comment:
I'm not sure if I can follow why this is in the `SymfonyRuntime` instead of the `ConsoleApplicationRunner`. Is a runner meant to only run the `$application` and should the runtime do all bootstrap?

Reply:
The decider is that anything that relates to external state (aka global state) should happen in the runtime. This provides flexibility by enabling decorating or subclass runtime to override just what they need for their specialization.

Reply:
Alright, that makes sense. Thanks for the clear description!